Improve accuracy of EventBuffer estimatedSerializedBytes

oliverhsu77 opened the pull request:
Inaccurate `EventBuffer. estimatedSerializedBytes` are causing the 6 MB add events payload limit to be exceeded.

Improve `EventBuffer. estimatedSerializedBytes` with the following changes:
1. More accurate per event overhead constant
2. Include server attributes in the estimatedSerializedBytes
3. Include String escaping for JSON strings in the estimatedSerializedBytes.

These changes indicate that the estimates are within 5% accuracy of the actual serialized payload.

Best to review by commit since I refactored `EventBuffer` to be its own top level class.  The crux of the changes are in the `Additional improvements for EventBuffer estimatedSerializedBytes` commit.  It's probably easiest to start from there and work backwards.  This commit also changed some of the changes in the first commit.  I typically would have rebased and combined the 1st and 3rd commits but refactoring the `EventBuffer` class made this difficult.

czerwingithub commented on src/main/java/com/scalyr/integrations/kafka/Event.java:
```diff
@@ -169,18 +180,49 @@ public int estimatedSerializedBytes() {
+      }
+      sb.setLength(0);
+
+      jsonStringEncoder.quoteAsString(s, sb);
```
Do you know what the performance impact of this is?  For json logs, this seems like it's going double the amount of encoding work that needs to be done.. It's not clear to me what this does for the GC pressure as well.  Since it is going directly into a StringBuilder, maybe it's not so bad?  But, it does worry me.

Why not just use a simple hueristic in its place?  I suspect most of the extra characters come from escaping double quotes, newlines and carriage returns.  Why not just count those characters and get an estimate?

Kami replied:
Yeah my main concern also is a potential performance impact of this code.

Another question also is if and how this will work in real-life (e.g. the user could be sending us already pre-escaped JSON or similar (seen that before) and we wouldn't even go in this code path - although I guess if it's already pre escaped we probably wouldn't try to escape it again?), but I guess only way to check and test that is to get access to the raw payloads once we exceed the limit and audit them.

Perhaps we could gather some real life JSON payloads and run your implementation and the one Steven proposed on it and see if the estimates are close enough and if they are, use a simpler (and probably faster) version.

oliverhsu77 replied:
Just realized in the middle of the night that re-using `StringBuilder` as a `static` will have multithreading issues.  I did this for performance to avoid allocating new Strings/StringBuffer.  We are single threaded in a single task (which is what I was thinking about yesterday), but in a single JVM we have multiple tasks on separate threads.  Will address this and performance questions in the morning.

oliverhsu77 replied:
I think we should remove the the `StringBuilder` and `JsonStringEncoder`.  Possible alternatives are 1) Get a rough count estimate by string length + counting escapable characters. or 2) Don't try to estimate String escapable characters

oliverhsu77 commented on src/main/java/com/scalyr/integrations/kafka/Event.java:
```diff
@@ -169,18 +180,49 @@ public int estimatedSerializedBytes() {
+    if (s.charAt(0) == '{' && s.charAt(s.length() - 1) == '}') {
+      StringBuilder sb = jsonStringBuilder.get();
+      if (sb == null) {
+        sb = new StringBuilder(s.length() + 20);
```
This code is now removed.  The original reason for doing this is that the  escaped strings are longer than the original String.  The StringBuilder would need to pretty quickly grow the underlying array resulting in a copy.  20 was an somewhat arbitrary estimate of additional String length.

oliverhsu77 commented:
Addressed review comments and made the following changes:
1. Used heuristic of counting JSON escapable characters to estimate serialized JSON size
2. Merged in Tomaz changes and resolved merge conflicts.  Those changes were using `sun.misc.IOUtils`, which wouldn't compile on my system with a deprecation warning.  Replaced `sun.misc.IOUtils` with Guava's `ByteStreams`.

Kami commented:
> Merged in Tomaz changes and resolved merge conflicts. Those changes were using sun.misc.IOUtils, which wouldn't compile on my system with a deprecation warning. Replaced sun.misc.IOUtils with Guava's ByteStreams.

Thanks!

Kami commented on src/main/java/com/scalyr/integrations/kafka/Event.java:
```diff
@@ -169,18 +187,48 @@ public int estimatedSerializedBytes() {
+      return s.length();
+    }
+
+    if (s.charAt(0) == '{' && s.charAt(s.length() - 1) == '}') {
```
Would this still work if JSON is already escaped? Or would that cause issues on the server side so it's not something we need to handle.

IIRC, I've seen some prescaped events which appeared to be correctly handled on the server side, but not sure if it matters or it's relevant here.

oliverhsu77 replied:
We're estimating payload size and not mutating the original payload.  Pre-escaped json throws off the count some.  e.g.
`{"req": "{\"jobId\":\"5eef5d97dc0e82000a8b6a15\"}` gets serialized in the add events payload to `{\"req\": \"{\\\"jobId\\\":\\\"5eef5d97dc0e82000a8b6a15\\\"}` resulting in a 11 character increase.  Notice it doesn't double-escape the `\"` escape.  If we just count escapable characters, we get 22 instead of 11.

It shouldn't have any effect on the server.  The server doesn't undo pre-escaped JSON.  i.e. It displays `{"req": "{\"jobId\":\"5eef5d97dc0e82000a8b6a15\"}` as `{"req": "{\"jobId\":\"5eef5d97dc0e82000a8b6a15\"}`.

Kami commented on src/main/java/com/scalyr/integrations/kafka/Event.java:
```diff
@@ -169,18 +187,48 @@ public int estimatedSerializedBytes() {
+    return s.length();
+  }
+
+  private int countJsonEscapedCharacters(String s) {
```
Do we have any idea on how much overhead this adds for JSON objects of various sizes?

If you have some time, I think it would be good to add a test case which times this function for a large JSON object, just so we have some idea on the overhead / timing involved.

oliverhsu77 replied:
I tested this with a ~15K JSON payload from a customer.  I ran `countJsonEscapedCharacters` 100_000 times.  It got an average of 0.19 milliseconds per ~15K JSON.  ~0.2 milliseconds seems like reasonable overhead.
